packages/itwinui-react/src/core/Popover/Popover.tsx

```diff
@@ -215,19 +221,31 @@ export const usePopover = (options: PopoverOptions & PopoverInternalProps) => {
     ),
   });
 
+  const interactionsEnabledStates = React.useMemo(
```

Comment:
nit: this intermediate variable looks unnecessary

Reply:
I felt it would be safe to use this variable to clearly specify which interactions are enabled. Then wherever we need to run some logic conditionally on whether the interaction is enabled/disabled, we can use this variable instead of inlining an `if` condition (E.g. [1](https://github.com/iTwin/iTwinUI/pull/2021/files/10c41d4b745f0f636c2db9e5f731754a7559e364..a5d9f8f1eb6afbabc9c45f1a9d846c18bf66c6aa#diff-6dd66c2b86bfd08956a740b1aa4db0d9b0c5c5ae871748791e4ac6eb06ff6af9R247-R250), [2](https://github.com/iTwin/iTwinUI/pull/2021/files/10c41d4b745f0f636c2db9e5f731754a7559e364..a5d9f8f1eb6afbabc9c45f1a9d846c18bf66c6aa#diff-eb1d53cfaf5ebd716d4feb107991cc2f5775bf2d4e9cc83b9de60e2206dea595R233)).

That way, if the logic of when a variable is enabled is changed (e.g. `someInteraction === true` vs. just `someInteractions != null`), we can just change it in one place.

Moreover, since this was related to the private `usePopover` hook, I felt this approach would be alright.

Reply:
So if it's just about using the default interactions, can't you return `mergedInteractions` directly? This intermediate variable does literally nothing extra, as far as I can tell.

Reply:
`mergedInteractions` is the `interactions` from the user merged with our default interactions. However, to know if an interaction is really enabled, we have to do something like `mergedInteractions.click === true || (typeof mergedInteractions.click === 'object' && mergedInteractions.click.enabled !== false)`.

We can write this condition each time we want to check if an interaction is enabled. However, to make it easy, `interactionsEnabledStates` runs these conditions and exposes an object where the key is the interaction, and the value is a boolean that represents whether the interaction is enabled or not.

This is why I think `interactionsEnabledStates` is helpful.

---

I realized the previous if-conditions were insufficient. I thus updated those if-conditions to also check for the `.enabled` (https://github.com/iTwin/iTwinUI/pull/2021/commits/3c4289471f31e2f67b8f2a9b6a2d9e1ea3823001).

Reply:
> I don't understand why it needs to be this complicated.

Yes, it's unfortunately complicated, but I tried to keep it as simple as possible. I felt it would be better to keep this slight complexity in a helper object instead of keeping it inline each time we need to check if an interaction is enabled since I felt the latter is error prone.

> Where do you even have a need for such checks??

* This is needed to run some workaround logic in parallel with FloatingUI's `onClick`. This logic only needs to run when the click interaction is enabled ([code](https://github.com/iTwin/iTwinUI/pull/2021/files/10c41d4b745f0f636c2db9e5f731754a7559e364..cffb1d0595c96bda2af2a7f8476e56f6c65c89c2#diff-eb1d53cfaf5ebd716d4feb107991cc2f5775bf2d4e9cc83b9de60e2206dea595R262-R265)).
* Updating `listRef` only when the listNavigation interaction is enabled. This is helpful because useFocusableNodes is, like you said, an expansive operation ([code](https://github.com/iTwin/iTwinUI/pull/2021/files/10c41d4b745f0f636c2db9e5f731754a7559e364..6b0bc399d7e1a07f4fa65229e31543e6e508b88e#diff-eb1d53cfaf5ebd716d4feb107991cc2f5775bf2d4e9cc83b9de60e2206dea595R201-R204)).

Reply:
Your links are broken. Can you permalink to the actual code in your branches instead of the PR diff?

Reply:
Ohh, I see. Looks like that's because `Menu.tsx` is collapsed by default in the diff view. Updated the links to permalinks.

Reply:
> This is needed to run some workaround logic in parallel with FloatingUI's `onClick`. This logic only needs to run when the click interaction is enabled ([code](https://github.com/iTwin/iTwinUI/blob/cffb1d0595c96bda2af2a7f8476e56f6c65c89c2/packages/itwinui-react/src/core/Menu/Menu.tsx#L262-L265)).

If this is a workaround, can we expand the code comment to indicate what exactly it is working around? We should try to remove it in the future.

> Updating `listRef` only when the listNavigation interaction is enabled. This is helpful because useFocusableNodes is, like you said, an expansive operation ([code](https://github.com/iTwin/iTwinUI/blob/6b0bc399d7e1a07f4fa65229e31543e6e508b88e/packages/itwinui-react/src/core/Menu/Menu.tsx#L201-L204)).

Isn't `listNavigation` always going to be enabled in `Menu`? And if it's an expensive operation, the correct solution is to only update the focusable nodes when necessary. Putting it in a `useEffect` with a large dependency array is probably not a good idea.

Reply:
> > If this is a workaround, can we expand the code comment to indicate what exactly it is working around? We should try to remove it in the future.
> 
> To confirm, do you mean in addition to this ([code](https://github.com/iTwin/iTwinUI/blob/6b0bc399d7e1a07f4fa65229e31543e6e508b88e/packages/itwinui-react/src/core/Menu/Menu.tsx#L277-L281))?

I didn't realize that comment was applicable for referring to this case, since it comes after the early return. Maybe the two separate comments should be combined.

Regardless, I was looking into this issue and realized you might be able to avoid the workaround using [`ignoreMouse: true`](https://floating-ui.com/docs/useclick#ignoremouse).

---

> > Isn't `listNavigation` always going to be enabled in `Menu`? And if it's an expensive operation, the correct solution is to only update the focusable nodes when necessary. Putting it in a `useEffect` with a large dependency array is probably not a good idea.
> 
> I'm not quite sure what is the best time/condition to update `listNavigationProps.listRef`. Is there a good way to know when the focusable nodes have changed?

There is no nice way but you can use a [mutation observer](https://developer.mozilla.org/en-US/docs/Web/API/MutationObserver/MutationObserver) to know when the child elements change.

In fact this could be abstracted away into a hook that always returns the latest focusable nodes, which you can then filter according to your needs.

```js
const allFocusableNodes = useFocusableNodes(rootNode);
const focusableNodes = React.useMemo(() => {
  return allFocusableNodes.filter(
    (i) => !focusableItems.some((p) => p.contains(i.parentElement)),
  );
}, [allFocusableNodes]);
```